File: app/transform/backend/src/main/java/com/carddemo/service/ReportService.java
package com.carddemo.service;

import com.carddemo.dto.AccountReportDto;
import com.carddemo.dto.CardReportDto;
import com.carddemo.dto.TransactionReportDto;

/**
 * Service interface for report generation
 * Provides methods to generate various reports on accounts, transactions, and cards
 */
public interface ReportService {

    /**
     * Generate account report with aggregated account data
     * @param status Optional filter by account status ('Y' for active, 'N' for inactive, null for all)
     * @param startDate Optional filter by open date (YYYY-MM-DD format)
     * @param endDate Optional filter by open date (YYYY-MM-DD format)
     * @return AccountReportDto containing aggregated account information
     */
    AccountReportDto generateAccountReport(String status, String startDate, String endDate);

    /**
     * Generate transaction report with aggregated transaction data
     * @param startDate Optional filter by transaction date (YYYY-MM-DD format)
     * @param endDate Optional filter by transaction date (YYYY-MM-DD format)
     * @param transactionType Optional filter by transaction type
     * @return TransactionReportDto containing aggregated transaction information
     */
    TransactionReportDto generateTransactionReport(String startDate, String endDate, String transactionType);

    /**
     * Generate card report with aggregated card data
     * @param status Optional filter by card status
     * @param expiringBefore Optional filter for cards expiring before this date (YYYY-MM-DD format)
     * @return CardReportDto containing aggregated card information
     */
    CardReportDto generateCardReport(String status, String expiringBefore);
}

File: app/transform/frontend/src/services/reportService.js
import api from './api'

/**
 * Report Service
 * Handles report generation operations
 */
export default {
  /**
   * Generate account report
   * @param {object} filters - Report filters
   * @param {string} filters.status - Account status filter ('Y' or 'N')
   * @param {string} filters.startDate - Start date filter (YYYY-MM-DD)
   * @param {string} filters.endDate - End date filter (YYYY-MM-DD)
   * @returns {Promise} Account report data
   */
  async generateAccountReport(filters = {}) {
    const params = new URLSearchParams()
    if (filters.status) params.append('status', filters.status)
    if (filters.startDate) params.append('startDate', filters.startDate)
    if (filters.endDate) params.append('endDate', filters.endDate)
    
    const response = await api.get(`/reports/accounts?${params.toString()}`)
    return response.data
  },

  /**
   * Generate transaction report
   * @param {object} filters - Report filters
   * @param {string} filters.startDate - Start date filter (YYYY-MM-DD)
   * @param {string} filters.endDate - End date filter (YYYY-MM-DD)
   * @param {string} filters.type - Transaction type filter
   * @returns {Promise} Transaction report data
   */
  async generateTransactionReport(filters = {}) {
    const params = new URLSearchParams()
    if (filters.startDate) params.append('startDate', filters.startDate)
    if (filters.endDate) params.append('endDate', filters.endDate)
    if (filters.type) params.append('type', filters.type)
    
    const response = await api.get(`/reports/transactions?${params.toString()}`)
    return response.data
  },

  /**
   * Generate card report
   * @param {object} filters - Report filters
   * @param {string} filters.status - Card status filter
   * @param {string} filters.expiringBefore - Expiring before date (YYYY-MM-DD)
   * @returns {Promise} Card report data
   */
  async generateCardReport(filters = {}) {
    const params = new URLSearchParams()
    if (filters.status) params.append('status', filters.status)
    if (filters.expiringBefore) params.append('expiringBefore', filters.expiringBefore)
    
    const response = await api.get(`/reports/cards?${params.toString()}`)
    return response.data
  }
}
